### ISSUE TITLE
Workflow stuck in running state when using activeDeadlineSeconds on template level

### ISSUE THREAD

**Issue by n1klasD** [argoproj/argo-workflows#12329]
### Pre-requisites

- [X] I have double-checked my configuration
- [X] I can confirm the issues exists when I tested with `:latest`
- [ ] I'd like to contribute the fix myself (see [contributing guide](https://github.com/argoproj/argo-workflows/blob/main/docs/CONTRIBUTING.md))

### What happened/what did you expect to happen?

When `activeDeadlineSeconds` is set on the template level of a workflow as described [here](https://argoproj.github.io/argo-workflows/walk-through/timeouts/), it does not transition to the `failed` state after the specified time. The workflow is stuck in the `running` state and it´s pods are not cleaned up. This seems to happen, when the deadline of `activeDeadlineSeconds` is exceeded while the workflow pods are still in `pending` state. To reproduce this using the provided workflow, you might need to adjust `activeDeadlineSeconds`, so that the deadline is exceeded while the fan-out steps are still in `pending` state.

The provided workflow stays in the running state, after the deadline is exceeded (it was manually stopped after two hours). 
The fan-out steps stay in the `Pending` state, but display "Pod was active on the node longer than the specified deadline":

```
Name:                active-deadline-fanout-template-level-pprjp
Namespace:           XXXX
ServiceAccount:      unset
Status:              Running
Conditions:
 PodRunning          False
Created:             Thu Dec 07 11:44:48 +0100 (2 hours ago)
Started:             Thu Dec 07 11:44:48 +0100 (2 hours ago)
Duration:            2 hours 54 minutes
Progress:            0/4

STEP                                            TEMPLATE    PODNAME                                                      DURATION  MESSAGE
 ● active-deadline-fanout-template-level-pprjp  entrypoint

 └─┬─◷ fanout(0:1)                              echo        active-deadline-fanout-template-level-pprjp-echo-628187050   2h        Pod was active on the node longer than the specified deadline
   ├─◷ fanout(1:2)                              echo        active-deadline-fanout-template-level-pprjp-echo-2925824112  2h        Pod was active on the node longer than the specified deadline
   ├─◷ fanout(2:3)                              echo        active-deadline-fanout-template-level-pprjp-echo-2656767666  2h        Pod was active on the node longer than the specified deadline
   └─◷ fanout(3:4)                              echo        active-deadline-fanout-template-level-pprjp-echo-606271796   2h        Pod was active on the node longer than the specified deadline
```

The workflow pods are not cleaned up and remain in the cluster. When looking at the workflow pods using `kubectl describe pods`, I noticed that they have the `Failed` status, but the `init` and `main` containers are stuck in the `Waiting` state with the reason `PodInitializing`:

```yaml
Name:                 active-deadline-fanout-template-level-pprjp-echo-2925824112
Namespace:            argo-dev
Priority:             1000
Priority Class Name:  low-priority-nonpreempting
Service Account:      default
Node:                 ####### redacted #######
Start Time:           Thu, 07 Dec 2023 11:44:48 +0100
Labels:               workflows.argoproj.io/completed=false
                      workflows.argoproj.io/workflow=active-deadline-fanout-template-level-pprjp
Annotations:          kubectl.kubernetes.io/default-container: main
                      sidecar.istio.io/inject: false
                      workflows.argoproj.io/node-id: active-deadline-fanout-template-level-pprjp-2925824112
                      workflows.argoproj.io/node-name: active-deadline-fanout-template-level-pprjp[0].fanout(1:2)
Status:               Failed
Reason:               DeadlineExceeded
Message:              Pod was active on the node longer than the specified deadline
IP:                   ### Redacted ###
IPs:
  IP:           ### Redacted ###
Controlled By:  Workflow/active-deadline-fanout-template-level-pprjp
Init Containers:
  init:
    Container ID:  containerd://a74281dd012d09a0bf9a93580fdf7bdba91bc28308a152cc31f007f1b599ab21
    Image:         quay.io/argoproj/argoexec:v3.5.2
    Image ID:      quay.io/argoproj/argoexec@sha256:407d10ce415c2ad6a7e8e72f120095d9f25bf6b253a6d18b59e36f7fde0e4c3a
    Port:          <none>
    Host Port:     <none>
    Command:
      argoexec
      init
      --loglevel
      warning
      --log-format
      text
    State:          Terminated
      Reason:       Completed
      Exit Code:    0
      Started:      Thu, 07 Dec 2023 11:44:50 +0100
      Finished:     Thu, 07 Dec 2023 11:45:16 +0100
    Ready:          True
    Restart Count:  0
    Limits:
      cpu:     10m
      memory:  124Mi
    Requests:
      cpu:     5m
      memory:  64Mi
    Environment:
      ARGO_POD_NAME:                      active-deadline-fanout-template-level-pprjp-echo-2925824112 (v1:metadata.name)
      ARGO_POD_UID:                        (v1:metadata.uid)
      GODEBUG:                            x509ignoreCN=0
      ARGO_WORKFLOW_NAME:                 active-deadline-fanout-template-level-pprjp
      ARGO_WORKFLOW_UID:                  f976104a-ddb9-474d-acd0-1e8790e4c1c6
      TRANSIENT_ERROR_PATTERN:            ### Redacted ###
      ARGO_CONTAINER_NAME:                init
      ARGO_TEMPLATE:                      {"name":"echo","inputs":{"parameters":[{"name":"item","value":"2"}]},"outputs":{},"metadata":{},"container":{"name":"","image":"alpine:latest","command":["sh","-c"],"args":["echo","workflow number 2"],"resources":{}},"activeDeadlineSeconds":20}
      ARGO_NODE_ID:                       active-deadline-fanout-template-level-pprjp-2925824112
      ARGO_INCLUDE_SCRIPT_OUTPUT:         false
      ARGO_DEADLINE:                      2023-12-09T10:44:48Z
      ARGO_PROGRESS_FILE:                 /var/run/argo/progress
      ARGO_PROGRESS_PATCH_TICK_DURATION:  1m0s
      ARGO_PROGRESS_FILE_TICK_DURATION:   3s
    Mounts:
      /var/run/argo from var-run-argo (rw)
      /var/run/secrets/kubernetes.io/serviceaccount from kube-api-access-lzwjk (ro)
Containers:
  wait:
    Container ID:  
    Image:         quay.io/argoproj/argoexec:v3.5.2
    Image ID:      
    Port:          <none>
    Host Port:     <none>
    Command:
      argoexec
      wait
      --loglevel
      warning
      --log-format
      text
    State:          Waiting
      Reason:       PodInitializing
    Ready:          False
    Restart Count:  0
    Limits:
      cpu:     10m
      memory:  124Mi
    Requests:
      cpu:     5m
      memory:  64Mi
    Environment:
      ARGO_POD_NAME:                      active-deadline-fanout-template-level-pprjp-echo-2925824112 (v1:metadata.name)
      ARGO_POD_UID:                        (v1:metadata.uid)
      GODEBUG:                            x509ignoreCN=0
      ARGO_WORKFLOW_NAME:                 active-deadline-fanout-template-level-pprjp
      ARGO_WORKFLOW_UID:                  f976104a-ddb9-474d-acd0-1e8790e4c1c6
      TRANSIENT_ERROR_PATTERN:            ### Redacted ###
      ARGO_CONTAINER_NAME:                wait
      ARGO_TEMPLATE:                      {"name":"echo","inputs":{"parameters":[{"name":"item","value":"2"}]},"outputs":{},"metadata":{},"container":{"name":"","image":"alpine:latest","command":["sh","-c"],"args":["echo","workflow number 2"],"resources":{}},"activeDeadlineSeconds":20}
      ARGO_NODE_ID:                       active-deadline-fanout-template-level-pprjp-2925824112
      ARGO_INCLUDE_SCRIPT_OUTPUT:         false
      ARGO_DEADLINE:                      2023-12-09T10:44:48Z
      ARGO_PROGRESS_FILE:                 /var/run/argo/progress
      ARGO_PROGRESS_PATCH_TICK_DURATION:  1m0s
      ARGO_PROGRESS_FILE_TICK_DURATION:   3s
    Mounts:
      /tmp from tmp-dir-argo (rw,path="0")
      /var/run/argo from var-run-argo (rw)
      /var/run/secrets/kubernetes.io/serviceaccount from kube-api-access-lzwjk (ro)
  main:
    Container ID:  
    Image:         alpine:latest
    Image ID:      
    Port:          <none>
    Host Port:     <none>
    Command:
      /var/run/argo/argoexec
      emissary
      --loglevel
      warning
      --log-format
      text
      --
      sh
      -c
    Args:
      echo
      workflow number 2
    State:          Waiting
      Reason:       PodInitializing
    Ready:          False
    Restart Count:  0
    Environment:
      ARGO_CONTAINER_NAME:                main
      ARGO_TEMPLATE:                      {"name":"echo","inputs":{"parameters":[{"name":"item","value":"2"}]},"outputs":{},"metadata":{},"container":{"name":"","image":"alpine:latest","command":["sh","-c"],"args":["echo","workflow number 2"],"resources":{}},"activeDeadlineSeconds":20}
      ARGO_NODE_ID:                       active-deadline-fanout-template-level-pprjp-2925824112
      ARGO_INCLUDE_SCRIPT_OUTPUT:         false
      ARGO_DEADLINE:                      2023-12-09T10:44:48Z
      ARGO_PROGRESS_FILE:                 /var/run/argo/progress
      ARGO_PROGRESS_PATCH_TICK_DURATION:  1m0s
      ARGO_PROGRESS_FILE_TICK_DURATION:   3s
    Mounts:
      /var/run/argo from var-run-argo (rw)
      /var/run/secrets/kubernetes.io/serviceaccount from kube-api-access-lzwjk (ro)
Conditions:
  Type              Status
  Initialized       True 
  Ready             False 
  ContainersReady   False 
  PodScheduled      True
Events:                      <none>
```

When setting `activeDeadlineSeconds` on [workflow level](https://argoproj.github.io/argo-workflows/walk-through/timeouts/)  instead, it transitions the workflow to the failed state after the deadline is exceeded as expected. `activeDeadlineSeconds` seems to work correctly on workflow level, but not on template level. 

The correct behaviour can be reproduced using this template: 

```yaml
apiVersion: argoproj.io/v1alpha1
kind: Workflow
metadata:
  name: active-deadline-fanout-workflow-level
  namespace: argo-dev
spec:
  activeDeadlineSeconds: 20 # defined on workflow level
  entrypoint: entrypoint
  templates:
  - name: entrypoint
    steps:
    - - name: fanout
        template: echo
        arguments:
          parameters:
            - name: item
              value: "{{item}}"
        withItems:
          - 1
          - 2
          - 3
          - 4
  - name: echo
    inputs:
      parameters:
        - name: item
    container:
      image: alpine:latest
      command:
        - sh
        - '-c'
      args:
        - echo
        - 'workflow number {{inputs.parameters.item}}'
```

### Version

v3.5.2

### Paste a small workflow that reproduces the issue. We must be able to run the workflow; don't enter a workflows that uses private images.

```YAML
apiVersion: argoproj.io/v1alpha1
kind: Workflow
metadata:
  name: active-deadline-fanout-template-level
spec:
  entrypoint: entrypoint
  templates:
  - name: entrypoint
    steps:
    - - name: fanout
        template: echo
        arguments:
          parameters:
            - name: item
              value: "{{item}}"
        withItems:
          - 1
          - 2
          - 3
          - 4
  - name: echo
    inputs:
      parameters:
        - name: item
    container:
      image: alpine:latest
      command:
        - sh
        - '-c'
      args:
        - echo
        - 'workflow number {{inputs.parameters.item}}'
    activeDeadlineSeconds: 20 # defined on template level, not workflow level !
```

### Logs from the workflow controller

```text
-
```

### Logs from in your workflow's wait container

```text
-
```

**Comment by Joibel** [comment]
This may be addressed by #12196 (not yet in any release).

**Comment by shuangkun** [comment]
Yeah, I test it when merge 12196. It  works well.

**Comment by shuangkun** [comment]
Now released in 3.5.3， you can test it.

**Comment by agilgur5** [comment]
#12196 is only when the Workflow is terminated, but the scenario as written does not terminate the Workflow.

I think we may want to extend similar logic for deleting `Pending` pods when `activeDeadlineSeconds` is passed (at either template or Workflow level)
